### DISCUSSION TITLE
[@types/react] React.useMemo failing when type checking object arrays

### DISCUSSION THREAD

**Discussion by colemars** [DefinitelyTyped/DefinitelyTyped#61826] [Issues with a @types package] [answered]
[Typescript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AoAekrgHUkAbNXJOGCNgC1YBUBPMEgDKaKMDDwACgxR8A5jgCuAOwAmAGi7AAznB1wUcAO5IARtuAwkVGkc7A0nOHOAA3JLr4RFBuI5TKaKzsxmJWmtqcKEQGanAMqFDKcPyCImISAHTkNnAAmt5+3gyqWsoA1nAQmHrwwMkwnESsRrLaAFw2uXAAtHAAgvEQaCgwwBANHAnRyanCouLwdkgxyhCN9XJ+AXCmRCjl3X2DYDJBbBxIAB6CYiBIykuWTnPpi3DafI8oV5oBpZFosFuHAAKoIAAyuiMzyqjRW2iOAw+-1MECuFzg11uoAeT0acFUwEwmBWeKK4GACSgcFJo0URF01RSAnmGRgOWYym08ACAFErihwAkAGrRYAoUwJOAAXjgACIABKMBgcWjQEry8hc7QQBKZVVyAAU-MFwqQYrEkoSAEoctQUlNEskQNBWJLvPB4fEAnJFCg5EhNGgGA5Kh73HpkvKBULTh55VUafLaFEYPhdAA5JBGeXZB0AeXhUBh2iDziQ8B50SspVMfDgRFdrk2XA8rGYIHuj10-zbxg1pX0XgZcFOsgU3jU2XIMFZyLlAG9lEKkO0PjAxMo5ABfChwODaiY8lIeeByo022UAPjgi-IB4POt5ssQxBgmUUZYAskhXQAefoAG0AF1ryNS8bzgIDl1Xdd8AARgAJgAZnwTQV3uJD12QlCdxAzRQLtR9G0rBlkn-IlXGvXD-0oKjr3IHdOWPeAeDPJDX0gmVb3vEjn18OVkHQD8vyQX8AOAsCIKvHjoNg+4UPg3D0LgTCkGwuBcPwwiQOIx8iBgci4EotwaNQuiGKYoA)

```ts
type A = {name: string};
  
// no type errors
const Test = () => {
    const a = React.useMemo<A[]>(() => [{name: '123', name2: 123}], [])
    return <div>123</div>
}

// type error 
// Type '{ name3: string; name2: number; }' is not assignable to type 'A'.
// Object literal may only specify known properties, but 'name3' does not exist in type 'A'. Did you mean to write 'name'?
const Test2 = () => {
    const a = React.useMemo<A[]>(() => [{name3: '123', name2: 123}], [])
    return <div>123</div>
}
```

Seems it's only checking the first property?

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment, marked as answer]
[Return types of functions are covariant](https://dmitripavlutin.com/typescript-covariance-contravariance/#1-subtyping) which is why TypeScript does not complain if you return a subtype of `A`:

```tsx
type A = {name: string};

function create<T>(factory: () => T): T {
  return factory()
}

create<A[]>(() => [{name: 'sd', name2: 'sd'}])
```

[Playground Link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UDeA7AhgWwgLigZ2ACcBLFAcwF8BuAKBoDMBXFAY2GIHsUoXCI1gEADwAVAHwAKemjYdCIXBICUCMVBFLcI5DShQ+wRoW7TZ85TQp1e-QUJgBtALqTlqqA9SYcUAOR4AE18AGih0LAAmXH8giiclIA)

I don't think there's anything we can do about this and I'm also not sure we should since it would change how TypeScript usually works.
